### PR TITLE
Initial implementation of destructuring assignment

### PR DESCRIPTION
**Pull request by cmxog**
This should help writing complicated code that extracts values from
structs.

### CODE REVIEW

File: lib/interpreter.ml [hunk 1]
```diff
@@ -151,6 +152,19 @@ class interpreter
...
               errors#report `Error `ArgumentNumberMismatch () ;
               Void )
+      | DestructuringLet let_ ->
+          let args_scope =
+            List.map let_.destructuring_let ~f:(fun (name, new_name) ->
+                let expr =
+                  StructField (let_.destructuring_let_expr, name, HoleType)
+                in
+                (new_name, self#interpret_expr expr) )
+          in
```
Comment (p0lunin) [inline lib/interpreter.ml]: This is not how destructuring should work. in your case you execute destructuring_let_expr N times where N - number of fields. You should execute destructuring_let_expr first and then get necessary fields from the output value.

File: lib/lang_types.ml [hunk 2]
```diff
@@ -78,13 +78,19 @@ and value =
 and stmt =
   | If of if_
   | Let of (string * expr) list
+  | DestructuringLet of destructuring_let
```
Comment (p0lunin) [inline lib/lang_types.ml:81]: Are you sure we need 2 kinds of let? We can simple make one `Let` with first field as `pattern` where `type pattern = BindPattern of string | StructDefinition of (string * pattern) list`.

File: lib/lang_types.ml [hunk 3]
```diff
@@ -78,13 +78,19 @@ and value =
...
   | Return of expr
   | Break of stmt
   | Expr of expr
   | Block of stmt list
   | Switch of switch
   | Invalid
 
+and destructuring_let =
+  { destructuring_let : (string * string) list;
```
Comment (p0lunin) [inline lib/lang_types.ml:90]: This is not general as it should be. Fully-qualified destructuring should be able to destructuring inner structs too, so you need to add `pattern` type as I wrote above.

File: lib/lexer.mll [hunk 4]
```diff
@@ -44,6 +44,7 @@ let whitespace = [' ' '\t']+
 rule token = parse
  | whitespace { token lexbuf }
  | newline  { next_line lexbuf; token lexbuf }
+ | ".." { REST }
```
Comment (p0lunin) [inline lib/lexer.mll]: We name tokens by their characters but `REST` is a semantic name, not syntax name.

**Review by p0lunin — CHANGES_REQUESTED** [review]

File: lib/lang_types.ml (hunk 2 above)
Comment (cmxog) [reply lib/lang_types.ml:81]: I suggest keeping them separate for the time being in the name of not doing what can amount to a relatively big change through the codebase with the only benefit being unifying two constructors into one.

We can certainly do that, but perhaps later?

File: lib/lang_types.ml (hunk 3 above)
Comment (cmxog) [reply lib/lang_types.ml:90]: I agree. It should be more general. However, wouldn't the current, shallow destructuring be useful today for implementing std?

If so, I suggest we merge this PR as is and add deep destructuring of inner structs as a separate change.

Comment (p0lunin) [reply lib/lang_types.ml:90]: Ok, let's merge this and improve later.

**Review by p0lunin — APPROVED** [review]

**Pull request merged**